torch_xla/csrc/xla_graph_executor.cpp

```diff
@@ -610,28 +619,85 @@ XLAGraphExecutor::ExecuteComputationWithBarrier(
       << ". Maybe the entry get "
          "kicked out of the LRU cache";
 
+  SyncTensorCollection coll;
+  coll.device = device;
+
   // Create DataPlaceHolder that will get filled in async executions.
   std::vector<xla::Shape>* output_shapes =
       DeviceContextArena::Get()->GetOutputShapesByHash(hash);
   std::vector<torch::lazy::BackendDataPtr> placeholders;
-  placeholders.reserve(output_shapes->size());
-  for (const xla::Shape& shape : *output_shapes) {
-    torch::lazy::BackendDataPtr handle =
-        WrapXlaData(runtime::GetComputationClient()->CreateDataPlaceholder(
-            device.toString(), std::move(shape)));
-    placeholders.push_back(handle);
+  std::vector<XLATensor::ShardingSpecPtr> sharding_specs;
+
+  if (static_cast<XlaDeviceType>(device.type()) == XlaDeviceType::SPMD) {
+    sharding_specs =
+        std::vector<XLATensor::ShardingSpecPtr>(output_shapes->size());
+    static std::unordered_map<torch::lazy::hash_t,
+                              std::vector<XLATensor::ShardingSpecPtr>,
+                              torch::lazy::HashReducer>
+        output_sharding_hash;
```

Comment:
nit. maybe we could leave a few lines of comments explining what we are doing here (e.g., caching the output sharding propagation results to avoid recomputing it in the subsequent runs).

Reply:
will do